core/include/kernel/tcg.h

```diff
@@ -0,0 +1,164 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Linaro Limited
+ *
+ * This file refers the following TCG specification.
+ * TCG PC Client Platform Firmware Profile Specification
+ */
+
+#ifndef __KERNEL_TCG_H__
+#define __KERNEL_TCG_H__
+
+#include <tee_api_types.h>
+#include <tpm2.h>
+
+#define TPM2_EVENT_LOG_SIZE		4096
+
+/*
+ * SHA1 Event Log Entry Format
+ *
+ * @pcr_index:  PCRIndex event extended to
+ * @event_type: Type of event (see EFI specs)
+ * @digest:     Value extended into PCR index
+ * @event_size: Size of event
+ * @event:      Event data
+ */
+struct tcg_pcr_event {
+	uint32_t pcr_index;
+	uint32_t event_type;
+	uint8_t digest[TPM2_SHA1_DIGEST_SIZE];
+	uint32_t event_size;
+	uint8_t event[];
+};
+
+/*
+ * Crypto Agile Log Entry Format
+ *
+ * @pcr_index:	PCRIndex event extended to
+ * @event_type:	Type of event
+ * @digests:	List of digestsextended to PCR index
+ * @event_size: Size of the event data
+ * @event:	Event data
+ */
+struct tcg_pcr_event2 {
+	uint32_t pcr_index;
+	uint32_t event_type;
+	struct tpml_digest_values digests;
+	uint32_t event_size;
+	uint8_t event[];
+} __packed;
+
+#define TCG_EFI_SPEC_ID_EVENT_SIGNATURE_03 "Spec ID Event03"
+#define TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_MAJOR_TPM2 2
+#define TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_MINOR_TPM2 0
+#define TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_ERRATA_TPM2 2
+
+/*
+ *  struct TCG_EfiSpecIdEventAlgorithmSize - hashing algorithm information
+ *
+ *  @algorithm_id:	algorithm defined in enum tpm2_algorithms
+ *  @digest_size:	size of the algorithm
+ */
+struct tcg_efi_spec_id_event_algorithm_size {
+	uint16_t      algorithm_id;
+	uint16_t      digest_size;
+};
+
+/**
+ * struct TCG_EfiSpecIDEventStruct - content of the event log header
+ *
+ * @signature:                  signature, set to Spec ID Event03
+ * @platform_class:             class defined in TCG ACPI Specification
+ *                              Client  Common Header.
+ * @spec_version_minor:         minor version
+ * @spec_version_major:         major version
+ * @spec_errata:                major version
+ * @uintn_size:                 size of the efi_uintn_t fields used in various
+ *                              data structures used in this specification.
+ *                              0x01 indicates uint32_t and 0x02 indicates
+ *                              uint64_t
+ * @number_of_algorithms:       hashing algorithms used in this event log
+ * @digest_sizes:               array of number_of_algorithms pairs
+ *                              1st member defines the algorithm id
+ *                              2nd member defines the algorithm size
+ */
+struct tcg_efi_spec_id_event {
+	uint8_t signature[16];
+	uint32_t platform_class;
+	uint8_t spec_version_minor;
+	uint8_t spec_version_major;
+	uint8_t spec_errata;
+	uint8_t uintn_size;
+	uint32_t number_of_algorithms;
+	struct tcg_efi_spec_id_event_algorithm_size digest_sizes[];
+} __packed;
```

Comment:
Why is this packed?

Reply:
Frankly, @apalos used it as __packed in u-boot, so I reused it as it is over here, :-). TPM structures are quiet unaligned. Eventlog header has a variable array at it's end and spec ID event is placed after that in event log. So it's starting boundary may be unaligned.